src/utils/__tests__/changelog.test.ts

```diff
@@ -0,0 +1,504 @@
+  ],
+  [
+    'prepend with underlined when detected',
+    '# Changelog\n\n1.0.0\n-----\n\nthis is a test\n',
+    '# Changelog\n\n2.0.0\n-----\n\n- rewrote everything from scratch\n- with multiple lines\n\n1.0.0\n-----\n\nthis is a test\n',
+  ],
+  [
+    'prepend with consistent padding with the rest',
+    '# Changelog\n\n   ## 1.0.0\n\n   this is a test\n',
+    '# Changelog\n\n   ## 2.0.0\n\n   - rewrote everything from scratch\n   - with multiple lines\n\n   ## 1.0.0\n\n   this is a test\n',
+  ],
+  [
+    'prepend with consistent padding with the rest (underlined)',
+    '# Changelog\n\n   1.0.0\n-----\n\n   this is a test\n',
+    '# Changelog\n\n   2.0.0\n-----\n\n   - rewrote everything from scratch\n   - with multiple lines\n\n   1.0.0\n-----\n\n   this is a test\n',
+  ],
+])('prependChangeset should %s', (_testName, markdown, expected) => {
+  expect(
+    prependChangeset(markdown, {
+      body: '- rewrote everything from scratch\n- with multiple lines',
+      name: '2.0.0',
+    })
+  ).toEqual(expected);
+});
+
+describe('generateChangesetFromGit', () => {
+  let mockClient: jest.Mock;
+  const makeCommitResponse = (commits: Record<string, string>[]) =>
+    mockClient.mockResolvedValueOnce({
+      repository: Object.fromEntries(
+        commits.map(({ hash, pr, milestone }: Record<string, string>) => [
+          `C${hash}`,
+          {
+            associatedPullRequests: {
+              nodes: pr
+                ? [
+                    {
+                      number: pr,
+                      milestone: milestone ? { number: milestone } : null,
+                    },
+                  ]
+                : [],
+            },
+          },
+        ])
+      ),
+    });
+
+  const mockGetChangesSince = getChangesSince as jest.MockedFunction<
+    typeof getChangesSince
+  >;
+  const dummyGit = {} as SimpleGit;
+
+  beforeEach(() => {
+    jest.resetAllMocks();
+    mockClient = jest.fn();
+    (getGitHubClient as jest.MockedFunction<
+      typeof getGitHubClient
+      // @ts-ignore we only need to mock a subset
+    >).mockReturnValue({ graphql: mockClient });
+  });
+
+  it('should not fail on empty changeset', async () => {
+    mockGetChangesSince.mockResolvedValueOnce([]);
+    const changes = await generateChangesetFromGit(dummyGit, '1.0.0');
+    expect(changes).toBe('');
+  });
+
+  describe('standalone changes', () => {
+    it('should use short commit SHA for local commits', async () => {
+      mockGetChangesSince.mockResolvedValueOnce([
+        {
+          hash: 'abcdef1234567890',
+          message: 'Upgraded the kernel',
+          pr: null,
+        },
+      ]);
+      makeCommitResponse([]);
+      const changes = await generateChangesetFromGit(dummyGit, '1.0.0');
+      expect(changes).toMatchInlineSnapshot(`
+        "## Various fixes & improvements
+
+        - Upgraded the kernel (abcdef12)"
```

Comment:
Is it worth it to test the hash length (to be more than 7)?

Reply:
We already do that on line 313, right? That's where this value comes from.

Reply:
My comment was more about adding another check to prevent a regression be in the snapshots, but that's another thing 🙂.